### PR TITLE
GH-42193: [Java] Update dependency to maintain JUnit 5 only

### PR DESCRIPTION
**Pull request by llama90**
### Rationale for this change

We have completed migrating all unit test code from JUnit 4 to JUnit 5 for all Java modules. Now, we are removing the compatibility library `junit-vintage-engine` to maintain JUnit 5 only.

### What changes are included in this PR?

- Removing JUnit 4 dependencies
- Updating remaining `junit.framework.TestCase` imports
- Updating excludes to prevent JUnit 4 dependencies

### Are these changes tested?

Yes.

### Are there any user-facing changes?

No.

* GitHub Issue: #42193

### CODE REVIEW

**Comment by llama90** [comment]
@vibhatha Is it sufficient to just remove the `junit-vintage-engine` dependency?

Also, I encountered an error after merging my Flight Module commit as below. However, the current machine was not used to migrate the Flight Module. The machine has a tricky environment, so I cannot be sure if the error is occurring for everyone.

However, the CIs seem to be working well.

```bash
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.026 s <<< FAILURE! -- in org.apache.arrow.driver.jdbc.ITDriverJarValidation
[ERROR] org.apache.arrow.driver.jdbc.ITDriverJarValidation.validateShadedJar -- Time elapsed: 0.011 s <<< FAILURE!
org.opentest4j.AssertionFailedError: jar ==> expected: <Driver jar was not detected in the classpath> but was: <jar>
        at org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
        at org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
        at org.junit.jupiter.api.AssertEquals.failNotEqual(AssertEquals.java:197)
        at org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:182)
        at org.junit.jupiter.api.Assertions.assertEquals(Assertions.java:1156)
        at org.apache.arrow.driver.jdbc.ITDriverJarValidation.getJdbcJarFile(ITDriverJarValidation.java:72)
        at org.apache.arrow.driver.jdbc.ITDriverJarValidation.validateShadedJar(ITDriverJarValidation.java:84)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1597)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1597)

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   ITDriverJarValidation.validateShadedJar:84->getJdbcJarFile:72 jar ==> expected: <Driver jar was not detected in the classpath> but was: <jar>
[INFO] 
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0
```

**Comment by vibhatha** [comment]
I also got the same error locally when I was testing some other feature. We may need to take a look. Could you file an issue?

**Comment by vibhatha** [comment]
Also I am not sure if that would be the only dependency. Can we try intellisense and see if we still get the Junit4?

**Comment by llama90** [comment]
> Also I am not sure if that would be the only dependency. Can we try intellisense and see if we still get the Junit4?

We can check the dependencies using `mvn dependency`.

There are no JUnit 4.x dependencies.

So, I believe there is no need for further checks.

**Comment by llama90** [comment]
> I also got the same error locally when I was testing some other feature. We may need to take a look. Could you file an issue?

That's right. The issue appears intermittently. I'll summarize the details and create an issue.

**Comment by lidavidm** [comment]
I believe we have a way to configure disallowed dependencies for a stronger check

**Comment by vibhatha** [comment]
> maven-enforcer-plugin

I took a look, we already listed a few such dependencies. That's perfect. Let's use that.

File: java/pom.xml [hunk 1]
```diff
@@ -659,6 +646,9 @@ under the License.
                     <!-- Do not include annotation processors. Use the annotations-only artifacts -->
                     <exclude>org.immutables:value</exclude>
                     <exclude>org.checkerframework:checker</exclude>
+                    <!-- Exclude all JUnit 4 dependencies -->
+                    <exclude>junit:junit:4.*</exclude>
+                    <exclude>org.junit.vintage:junit-vintage-engine</exclude>
```
Comment (llama90) [inline java/pom.xml:651]: I updated it. Thank you for review!

**Comment by lidavidm** [comment]
@github-actions crossbow submit -g java

**Comment by llama90** [comment]
> I also got the same error locally when I was testing some other feature. We may need to take a look. Could you file an issue?

I created issue for [comment](https://github.com/apache/arrow/pull/42206#issuecomment-2177577528)

- https://github.com/apache/arrow/issues/42208

**Review by lidavidm — APPROVED** [review]

**Comment by lidavidm** [comment]
just making sure java-jars passes now

**Review by vibhatha — APPROVED** [review]

**Comment by vibhatha** [comment]
Did we re-run?

**Comment by lidavidm** [comment]
@vibhatha can someone take a look at the issue? I don't think this will go away on re-run

**Comment by vibhatha** [comment]
@lidavidm I think @kou resolved this issue here: https://github.com/apache/arrow/issues/42197
Probably need a rebase?

**Comment by llama90** [comment]
That was rebased about two hours ago. Perhaps we could try running it again?

**Comment by vibhatha** [comment]
@github-actions crossbow submit java-jars

**Comment by vibhatha** [comment]
@llama90 I think now we are failing at the issue we recently filed: https://github.com/apache/arrow/issues/42208

**Comment by lidavidm** [comment]
@llama90 if you want to rebase here I can re-run Crossbow

**Comment by llama90** [comment]
@lidavidm Done!

**Comment by lidavidm** [comment]
@github-actions crossbow submit -g java

**Review by lidavidm — APPROVED** [review]

**Pull request merged**